src/Microsoft.ML.TensorFlow/TensorFlow/TensorflowUtils.cs

```diff
@@ -15,65 +15,71 @@ namespace Microsoft.ML.Transforms.TensorFlow
 {
     public static class TensorFlowUtils
     {
-        public const string OpType = "OpType";
-        public const string InputOps = "InputOps";
+        /// <summary>
+        /// Key to access operator's type (a string) in <see cref="Schema.Column.Metadata"/>.
+        /// Its value describes the Tensorflow operator that produces this <see cref="Schema.Column"/>.
+        /// </summary>
+        public const string TensorflowOperatorTypeKind = "TensorflowOperatorType";
+        /// <summary>
+        /// Key to access upstream operators' names (a string array) in <see cref="Schema.Column.Metadata"/>.
+        /// Its value states operators that the associated <see cref="Schema.Column"/>'s generator depends on.
+        /// </summary>
+        public const string TensorflowUpstreamOperatorsKind = "TensorflowUpstreamOperators";
 
         internal static Schema GetModelSchema(IExceptionContext ectx, TFGraph graph, string opType = null)
         {
-            var res = new List<KeyValuePair<string, ColumnType>>();
-            var opTypeGetters = new List<MetadataUtils.MetadataGetter<ReadOnlyMemory<char>>>();
-            var inputOpsGetters = new List<MetadataUtils.MetadataGetter<VBuffer<ReadOnlyMemory<char>>>>();
-            var inputOpsLengths = new List<int>();
+            var schemaBuilder = new SchemaBuilder();
             foreach (var op in graph)
             {
                 if (opType != null && opType != op.OpType)
                     continue;
+
                 var tfType = op[0].OutputType;
+                // Determine element type in Tensorflow tensor. For example, a vector of floats may get NumberType.R4 here.
                 var mlType = Tf2MlNetTypeOrNull(tfType);
 
-                // If the type is not supported in ML.NET then we cannot represent it as a column in an ISchema.
+                // If the type is not supported in ML.NET then we cannot represent it as a column in an Schema.
                 // We also cannot output it with a TensorFlowTransform, so we skip it.
                 if (mlType == null)
                     continue;
 
-                var shape = graph.GetTensorShape(op[0]);
-                var shapeArray = shape.ToIntArray();
-
-                inputOpsLengths.Add(op.NumInputs);
-                MetadataUtils.MetadataGetter<VBuffer<ReadOnlyMemory<char>>> inputOpsGetter = null;
+                // Construct the final ML.NET type of a Tensorflow variable.
+                var shapeArray = graph.GetTensorShape(op[0]).ToIntArray();
+                var columnType = new VectorType(mlType);
+                if (Utils.Size(shapeArray) > 0 && shapeArray.Skip(1).All(x => x > 0))
+                    columnType = new VectorType(mlType, shapeArray[0] > 0 ? shapeArray : shapeArray.Skip(1).ToArray());
```

Comment:
>columnType [](start = 20, length = 10)

VectorType constructor throws exception when length of dimension is 0. I think this could happen when shapeArray = {1}. I see in previous code there is a special check for shapeArray.length of `1` and `0`. Can you please make sure if code works in this case.

Reply:
I don't see an exception in its original version. This code is equivalent to the old setting. If old code throws, it would throw as well.

Reply:
I see that you added a check `!(Utils.Size(shapeArray) == 1 && shapeArray[0] <= 0)` in commit 5. This will make `columnType == null` in case where shapeArray = {0} correct? 

What I meant to say that 
``` csharp
var columnType = Utils.Size(shapeArray) == 1 && shapeArray[0] == -1 ? new VectorType(mlType) :
                    Utils.Size(shapeArray) > 0 && shapeArray.Skip(1).All(x => x > 0) ?
                        new VectorType(mlType, shapeArray[0] > 0 ? shapeArray : shapeArray.Skip(1).ToArray())
                        : new VectorType(mlType);
```

is not equivalent to 

``` csharp
if (!(Utils.Size(shapeArray) == 1 && shapeArray[0] <= 0) &&
                    (Utils.Size(shapeArray) > 0 && shapeArray.Skip(1).All(x => x > 0)))
                    columnType = new VectorType(mlType, shapeArray[0] > 0 ? shapeArray : shapeArray.Skip(1).ToArray());
```

Test might be passing but there are cases when it will fail.

---
In reply to: [247230900](https://github.com/dotnet/machinelearning/pull/2123#discussion_r247230900) [](ancestors = 247230900)

Reply:
They are equivalent... That's why recursive if-else condition shouldn't be used anywhere. The only situation which is not `VectorType(mlType)` happens when the two conditions are both false. For all other cases, we get `VectorType(mlType)`.

---
In reply to: [247239512](https://github.com/dotnet/machinelearning/pull/2123#discussion_r247239512) [](ancestors = 247239512,247230900)

Reply:
OK, I see columtype is created at line 48 `var columnType = new VectorType(mlType);` with that its worsk now after 4th commit. Thanks for the explanation.

---
In reply to: [247259950](https://github.com/dotnet/machinelearning/pull/2123#discussion_r247259950) [](ancestors = 247259950,247239512,247230900)